### PR TITLE
WIP: add Diversity is All You Need implementation

### PR DESCRIPTION
**Pull request by kinalmehta**
## Description

Adds implementation of **Diversity is All You Need** paper. It is an unsupervised option learning framework which can later be used for transfer learning.
### To-Do
- [x] Implement unsupervised skill learning
- [ ] implement model saving and loading logic
- [ ] use the pre-trained model to train on a task from paper to benchmark
- [ ] add documentation and benchmarks

## Types of changes

- [ ] Bug fix
- [ ] New feature
- [x] New algorithm
- [ ] Documentation

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/vwxyzjn/cleanrl/blob/master/CONTRIBUTING.md) guide (**required**).
- [x] I have ensured `pre-commit run --all-files` passes (**required**).
- [ ] I have updated the documentation and previewed the changes via `mkdocs serve`.
- [ ] I have updated the tests accordingly (if applicable).

If you are adding new algorithms or your change could result in performance difference, you may need to (re-)run tracked experiments. See https://github.com/vwxyzjn/cleanrl/pull/137 as an example PR. 
- [ ] I have contacted [vwxyzjn](https://github.com/vwxyzjn) to obtain access to the [openrlbenchmark W&B team](https://wandb.ai/openrlbenchmark) (**required**).
- [ ] I have tracked applicable experiments in [openrlbenchmark/cleanrl](https://wandb.ai/openrlbenchmark/cleanrl) with `--capture-video` flag toggled on (**required**).
- [ ] I have added additional documentation and previewed the changes via `mkdocs serve`.
    - [ ] I have explained note-worthy implementation details.
    - [ ] I have explained the logged metrics.
    - [ ] I have added links to the original paper and related papers (if applicable).
    - [ ] I have added links to the PR related to the algorithm.
    - [ ] I have created a table comparing my results against those from reputable sources (i.e., the original paper or other reference implementation).
    - [ ] I have added the learning curves (in PNG format with `width=500` and `height=300`).
    - [ ] I have added links to the tracked experiments.
    - [ ] I have updated the overview sections at the [docs](https://docs.cleanrl.dev/rl-algorithms/overview/) and the [repo](https://github.com/vwxyzjn/cleanrl#overview)
- [ ] I have updated the tests accordingly (if applicable).

### CODE REVIEW

File: cleanrl/diayn_sac_continuous_action.py [hunk 1]
```diff
@@ -0,0 +1,636 @@
...
+        help="automatic tuning of the entropy coefficient")
+    parser.add_argument("--num-skills", type=int, default=50)
+    parser.add_argument("--discriminator_lr", type=float, default=1e-3,
+        help="the learning rate of the discriminator network")
+    parser.add_argument("--best-skill", type=int, default=None,
+        help="the index of the best skill (< num-skills)")
+    group = parser.add_mutually_exclusive_group()
+    group.add_argument("--learn-skills", action='store_true', default=False)
+    group.add_argument("--use-skills", action='store_true', default=False)
+    group.add_argument("--evaluate-skills", action='store_true', default=False)
```
Comment (vwxyzjn) [inline cleanrl/diayn_sac_continuous_action.py:80]: Please use the following configuration for bool:

```
    parser.add_argument("--autotune", type=lambda x:bool(strtobool(x)), default=False, nargs="?", const=True,
        help="automatic tuning of the entropy coefficient")
```

File: cleanrl/diayn_sac_continuous_action.py [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+#         super().__init__()
+#         self.fc1 = nn.Linear(np.array(env.single_observation_space.shape).prod(), 256)
+#         self.fc2 = nn.Linear(256, 256)
+#         self.fc3 = nn.Linear(256, num_skills)
+
+#     def forward(self, x):
+#         x = F.relu(self.fc1(x))
+#         x = F.relu(self.fc2(x))
+#         x = self.fc3(x)
+#         return OneHotCategorical(logits = x)
```
Comment (vwxyzjn) [inline cleanrl/diayn_sac_continuous_action.py:133]: This should go to docs, under the implementation details section.

File: cleanrl/diayn_sac_continuous_action.py [hunk 3]
```diff
@@ -0,0 +1,636 @@
...
+
+
+def aug_obs_z(obs, skill_one_hot):
+    obs = np.asarray(obs)
+    aug_obs = np.hstack((obs, skill_one_hot))
+    return aug_obs
+
+
+def split_aug_obs(aug_obs, num_skills):
+    assert type(aug_obs) in [torch.Tensor, np.ndarray] and type(num_skills) is int, "invalid input type"
```
Comment (vwxyzjn) [inline cleanrl/diayn_sac_continuous_action.py:197]: The check may not be needed for simplicity. Otherwise we may also need a check for `aug_obs_z`.

File: cleanrl/diayn_sac_continuous_action.py [hunk 4]
```diff
@@ -0,0 +1,636 @@
...
+    obs, skill_one_hot = aug_obs[:, :-num_skills], aug_obs[:, -num_skills:]
+    return obs, skill_one_hot
+
+
+class DIAYN:
+    def __init__(self, args, run_name=None, device=torch.device("cpu")):
+
+        self.args = args
+        self.device = device
```
Comment (vwxyzjn) [inline cleanrl/diayn_sac_continuous_action.py:206]: Please use the standard single-file implementation format in place of classes.

File: cleanrl/diayn_sac_continuous_action.py [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+            best_skill = np.argmax(scores)
+
+        # initializing the networks with best skill and the fine-tuning on it
+        one_hot_z = np.zeros((1, self.args.num_skills), dtype=np.float32)
+        one_hot_z[0, best_skill] = 1
+
+        # TRY NOT TO MODIFY: start the game
+        obs = self.envs.reset()
+        z_aug_obs = aug_obs_z(obs, one_hot_z)
+        for global_step in range(self.args.total_timesteps):
```
Comment (vwxyzjn) [inline cleanrl/diayn_sac_continuous_action.py:437]: What is the difference between `learn_skills` and `use_skills`? Why are both of them going over `1000000` steps?

File: cleanrl/diayn_sac_continuous_action.py [hunk 6]
```diff
@@ -0,0 +1,636 @@
...
+                    self.writer.add_scalar("losses/qf2_loss", qf2_loss.item(), global_step)
+                    self.writer.add_scalar("losses/qf_loss", qf_loss.item() / 2.0, global_step)
+                    self.writer.add_scalar("losses/actor_loss", actor_loss.item(), global_step)
+                    self.writer.add_scalar("losses/alpha", self.alpha, global_step)
+                    self.writer.add_scalar("losses/discriminator_loss", discriminator_loss.item(), global_step)
+                    print("SPS:", int(global_step / (time.time() - start_time)))
+                    self.writer.add_scalar("charts/SPS", int(global_step / (time.time() - start_time)), global_step)
+                    if self.args.autotune:
+                        self.writer.add_scalar("losses/alpha_loss", alpha_loss.item(), global_step)
+        return
```
Comment (vwxyzjn) [inline cleanrl/diayn_sac_continuous_action.py:543]: There is no need for return — it's implicit.

File: cleanrl/diayn_sac_continuous_action.py [hunk 7]
```diff
@@ -0,0 +1,636 @@
...
+        self.qf1.load_state_dict(models_info["qf1"])
+        self.qf1_target.load_state_dict(models_info["qf1_target"])
+        self.qf2.load_state_dict(models_info["qf2"])
+        self.qf2_target.load_state_dict(models_info["qf2_target"])
+        self.discriminator.load_state_dict(models_info["discriminator"])
+
+        # self.actor_optimizer.load_state_dict(models_info["actor_optimizer"])
+        # self.q_optimizer.load_state_dict(models_info["q_optimizer"])
+        # self.discriminator_optimizer.load_state_dict(models_info["discriminator_optimizer"])
+        return
```
Comment (vwxyzjn) [inline cleanrl/diayn_sac_continuous_action.py:603]: There is no need for return — it's implicit.

**Review by vwxyzjn — COMMENTED** [review]
Great work @kinalmehta! DIAYN looks like a pretty interesting paper. Some thoughts:

1. Have you run some preliminary experiments to see if you can replicate the results reported in the paper?
2. `learn_skills` and `use_skills` have a huge amount of duplicate code. If their purpose is to save and load models, consider the approach listed in https://docs.cleanrl.dev/advanced/resume-training/#resume-training_1.

File: cleanrl/diayn_sac_continuous_action.py (hunk 5 above)
Comment (kinalmehta) [reply cleanrl/diayn_sac_continuous_action.py:437]: `learn_skills` is the unsupervised skill learning phase, whereas `use_skills` is to fine-tune the trained model to optimize for the environment reward
